### PR TITLE
ebmc: check a liveness property using a given ranking function

### PR DESCRIPTION
**Pull request by kroening**
This adds a facility to prove a given liveness property using a given ranking function.

### CODE REVIEW

File: src/ebmc/ranking_function.cpp [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    throw 0;
+  }
+
+  for(auto &p : properties)
+  {
+    if(!p.is_disabled())
+      return p;
+  }
+
+  INVARIANT(false, "must not reach here");
```
Comment (tautschnig) [inline src/ebmc/ranking_function.cpp:115]: Use `UNREACHABLE;`?

File: src/ebmc/ranking_function.cpp [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+  // find the property
+  auto &property = find_property();
+
+  satcheckt satcheck{*message_handler};
+  boolbvt solver(ns, satcheck, *message_handler);
+
+  // *no* initial state, two time frames
+  unwind(*trans_expr, *message_handler, solver, 2, ns, false);
+
+  const exprt p = [&property]() -> exprt {
```
Comment (tautschnig) [inline src/ebmc/ranking_function.cpp:152]: Use `optionalt<unary_exprt>` as return type?

File: src/ebmc/ranking_function.cpp [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+};
+
+int do_ranking_function(
+  const cmdlinet &cmdline,
+  ui_message_handlert &ui_message_handler)
+{
+  return ranking_function_checkt(cmdline, ui_message_handler)();
+}
+
+exprt ranking_function_checkt::parse_ranking_function()
```
Comment (tautschnig) [inline src/ebmc/ranking_function.cpp:63]: How about using `optionalt<exprt>`?

File: src/ebmc/ranking_function.cpp [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    return 1;
+
+  // needs to have some numerical type
+  if(
+    ranking_function.type().id() != ID_signedbv &&
+    ranking_function.type().id() != ID_unsignedbv)
+  {
+    error() << "ranking function must have integral bitvector type" << eom;
+    return 1;
+  }
```
Comment (tautschnig) [inline src/ebmc/ranking_function.cpp:141]: Should this perhaps be part of `parse_ranking_function`?

**Review by tautschnig — APPROVED** [review]

File: src/ebmc/ranking_function.cpp (hunk 1 above)
Comment (kroening) [reply src/ebmc/ranking_function.cpp:115]: sure

File: src/ebmc/ranking_function.cpp (hunk 4 above)
Comment (kroening) [reply src/ebmc/ranking_function.cpp:141]: sure

**Pull request merged**